client/client.go

```diff
@@ -98,24 +111,20 @@ func (c *Client) prepareRequest(ctx context.Context, req *Request) (*http.Reques
 	return httpRequest, nil
 }
 
-func (c *Client) prepareResponse(resp *http.Response) (*Response, error) {
-	if resp.StatusCode < 200 || resp.StatusCode > 299 {
-		defer resp.Body.Close()
-
-		errorBody, readErr := ioutil.ReadAll(resp.Body)
-		if readErr != nil {
-			return nil, fmt.Errorf("got %d for %s: [no body]", resp.StatusCode, resp.Request.URL)
+func (c *Client) prepareResponse(ctx context.Context, resp *http.Response) (*Response, error) {
+	if c.problemDecoder != nil && resp.Header.Get(headers.ContentType) == "application/problem+json" {
```

Comment:
Yes, I should probably add a constant.

The nil check is needed when the user does not supply a ProblemDecoder, to avoid people using the client with Problem-able API:s and is expecting the old errors to be returned. Perhaps the statement will look better with a constant?

Reply:
What's the issue with always initializing a problemDecoder? Isn't the contentType header enough to determine which error to return?

Reply:
In my first iteration, I initialized the `problemDecoder` in `NewClient`. But the scenario I'm trying to avoid is something like,

* Service A is using Problems for reporting errors to their clients (i.e. they set the header)
* Client B is using our library to talk to Service A. They use the version before this change has been merged
* The old behavior was to ignore all contents of the body and only focus on the HTTP Status if there was a non-200 status returned by the server.
* If they bump their client library version they will start getting problems where things like, `errors.Is(err, client.ErrNotFound)` will start return false.

By making the problem decoding opt-in we avoid this issue. But as I remarked in the comment that tests this behavior, it should probably be changed upon the next major version release of this library.

Reply:
Reading between the lines: When we release v1 we **should** break backwards compatibilty. Forcing consumers to use problems and removing go-errors. I can't see how this problem will ever go away otherwise. 

As a user, I would prefer changing my error behavior in my code over having to worry about if I've configured this correctly to use problems or go-errors

Reply:
I think we will still have the HTTPError in v1, it's just the default behavior of the Client that will change. 

* Current version: do not return Problems unless the client/library has opt-in
* Next version: if a Problem is returned by the server return them by default (perhaps with an opt-out, if needed)

It all depends on the needs of the implementors of the client libraries such as go-hierarchy-client.

Reply:
Fair enough. Thanks for walking me through your thoughts.